### PR TITLE
Add GUI for working with demonstrations

### PR DESCRIPTION
**Pull request by petschekr**
Interfaces with the files in `hlpr_record_demonstration` to provide an easier way to record keyframes to a `.bag` file, load them from disk, inspect recordings, and play them back.

### CODE REVIEW

**Comment by petschekr** [comment]
Documentation: https://github.com/HLP-R/hlpr_documentation/wiki/Kinesthetic-Teaching-GUI

**Comment by vchu** [comment]
Tried to run the GUI and ran into an error

`RosPluginProvider.load(rqt_kinesthetic_teaching/HLP-R Kinesthetic Teaching) exception raised in __builtin__.__import__(rqt_kinesthetic_teaching.rqt_kinesthetic_teaching, [KinestheticTeaching]):
Traceback (most recent call last):
  File "/opt/ros/indigo/lib/python2.7/dist-packages/rqt_gui/ros_plugin_provider.py", line 77, in load
    module = __builtin__.__import__(attributes['module_name'], fromlist=[attributes['class_from_class_type']], level=0)
  File "/home/vchu/catkin_ws/src/HLP-R/hlpr_kinesthetic_teaching/rqt_kinesthetic_teaching/src/rqt_kinesthetic_teaching/rqt_kinesthetic_teaching.py", line 8, in <module>
    import kinesthetic_interaction_interface
ImportError: No module named kinesthetic_interaction_interface
`

Not sure if something isn't building? I tried a clean build and still encounter the same issue.

Basically - I don't know where I'm supposed to be getting a `kinesthetic_interaction_interface` from.  I do see a `keyframe_bag_interface`. Is that what it is supposed to be or is there a file missing?

**Comment by petschekr** [comment]
Hmm weird that Python didn't catch that error when I ran it. Try deleting line 8 (`import kinesthetic_interaction_interface`) in rqt_kinesthetic_teaching.py. It's an import that isn't needed anymore and points to a file that got renamed.

**Comment by vchu** [comment]
Yup that did it! Will continue testing the rest...

**Comment by vchu** [comment]
So I realized I don't have write access to your branch. You'll have to make the import change fix, commit it and then we can approve the PR

Edit: JK - realized I could push. Was trying with the wrong branch.

Going to wait on the PR still. Realized some commands are missing

**Comment by vchu** [comment]
Just ran through the code. I did the following sequence of events

(1) Started up the robot_bringup launch script (without basic kinesthetic mode, but with the playback and record services)
(2) started up the rqt kinesthetic GUI
(3) Saved to test2.bag
(4) Gave the following speech commands: (1) start here (then moved the arm) (2) close your hand (then moved the arm) (3) end here

I got the following segfault

> [WARN] [WallTime: 1497311511.424450] [164.902000] Waiting for speech service
[WARN] [WallTime: 1497311511.426339] [164.904000] Speech service loaded
[INFO] [WallTime: 1497311511.428665] [164.905000] No function in kinesthetic teaching for commands: SMALL_TALK, START_EXP, GREETING, HEAR_CHECK
[WARN] [WallTime: 1497311511.430679] [164.906000] Waiting for gravity compensation service
[WARN] [WallTime: 1497311511.433092] [164.907000] Gravity compenstation service loaded
[INFO] [WallTime: 1497311511.447643] [164.914000] Finished initializing Kinesthetic Interaction node. Service is currently set to: False
[WARN] [WallTime: 1497311511.468777] [164.931000] Waiting for record keyframe demo server to load
[WARN] [WallTime: 1497311511.597878] [165.008000] Record keyframe demo server loaded
Waiting for RQT kinesthetic interaction service
[INFO] [WallTime: 1497311511.605887] [165.010000] Kinesthetic Interaction set to: True
[Result] Num Poses Recorded: 3.000000
Segmentation fault (core dumped)

Also, just as a note - it would be nice to have a way to turn kinesthetic interaction mode to false as well. Maybe some checkbox of sorts that enable/disables kinesthetic mode rather than have it be specific to writing a bag file. This allows people to verbally give commands like (close your hand or release your arm) without having to be recording.

**Comment by vchu** [comment]
Is there a way to make rail_object_locator not dependent? If people want to do kteach without visual input, then it would be nice if the GUI doesn't crash.

System crashing. Need to reboot. Will continue testing this in the morning :)

**Comment by petschekr** [comment]
Oh yeah that wouldn't be too hard to implement. I'll tackle that in the morning.

**Comment by vchu** [comment]
Great thanks! Related to that - this needs to be done on the playback side as well. (i.e. if I have a bag file without object info, it should playback as well.) 

A couple of other notes

1. In general, if there is already a launch file for something, we want to include it rather than copy and paste it into new ones. So for object_location_engine.launch, you should probably change that launch file to call object_detector's launch file with whatever custom info you need. The launch file should probably also be called with a flag. It is possible that rail_object_detection could be running on a different machine.

2. The tutorial says to launch "start_robot_services". What it really should say is to launch "customize_start_robot_services.launch". This is an example of how to call "start_robot_services" with the correct flags. So ideally, each person or unique project should have their own version of "customize_start_robot_services.launch". This is also an example of how to call other launch files.

3. I tried reading/writing using both gui and voice interchangeably and it works great. Only suggestion is that the warning on the command line of "Recording saved but cannot be opened automatically because the end keyframe was written from a different thread. Please open the .bag file manually." This should either be a roswarn (so that it changes color) OR displayed somewhere on the GUI.

**Comment by petschekr** [comment]
@vchu Thanks for those notes. I believed I've fixed all of them.

**Comment by vchu** [comment]
1. Looks like if I use a mix of speech and GUI buttons I don't get any now warning to load the bag file again. 

2. When executing, I un-selected "Locate objects". However, I get an error that says

> [INFO] [WallTime: 1497622656.557555] [258.169000] Playing /home/vchu/data/kteach/test4.bag
Traceback (most recent call last):
  File "/home/vchu/catkin_ws/src/HLP-R/hlpr_kinesthetic_teaching/rqt_kinesthetic_teaching/src/rqt_kinesthetic_teaching/kinesthetic_teaching_widget.py", line 346, in playDemo
    keyframeBagInterface.play(location, zeroMarker, self.playDemoDone)
  File "/home/vchu/catkin_ws/src/HLP-R/hlpr_kinesthetic_teaching/rqt_kinesthetic_teaching/src/rqt_kinesthetic_teaching/keyframe_bag_interface.py", line 137, in play
    goal.zero_marker = ""
AttributeError: 'PlaybackKeyframeDemoGoal' object has no attribute 'zero_marker'

**Comment by petschekr** [comment]
@vchu I'll look into the first point. As for the second, you may have to `catkin_make` to rebuild the `.action` file that was changed when adding support for playing back relative to found objects.

**Comment by vchu** [comment]
Ah, yup that fixed it. 

1. Currently we get a rospy exception if I try to playback a bag file with the "Locate object" flag enabled and didn't have the object_locator service running. Is it possible to write this exception in a warning as well? Not high priority, since it seems to work as is still.

2. Just FYI - the wiki doesn't seem to be updated with the latest info. :)

3. I encountered an issue with the gripper not closing fast enough (or really the delay on the gripper KF) was too short during playback. Not related at all with the GUI I think? Might have to look deeper in the playback code. Have you encountered this at all?

**Comment by petschekr** [comment]
1. Just fixed that in 3181cdf
2. Yeah, I still need to add some information about the locate objects option and related stuff
3. I've noticed that too. It might be that the GUI is recording a frame partway through the gripper's closing or opening. This means MoveIt thinks the keyframe has been executed so it advances to the next keyframe. The gripper still closes all the way because the requested position of the gripper is correct.

**Comment by vchu** [comment]
@maxsvetlik 

FYI - merging in some changes for a HLP-R GUI. It shouldn't affect things things since I believe you use a different playback server entirely. Just FYI in case.
